viewer/src/utilities/IndexSet.ts

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright 2021 Cognite AS
+ */
+import assert from 'assert';
+
+import { NumericRange } from './NumericRange';
+
+type IndexNode = IntermediateIndexNode | LeafIndexNode;
+
+class IntermediateIndexNode {
+  readonly range: NumericRange;
+  readonly maxSubtreeDepth: number;
+  readonly left: IndexNode;
+  readonly right: IndexNode;
+  readonly count: number;
+
+  constructor(left: IndexNode, right: IndexNode) {
+    this.left = left;
+    this.right = right;
+
+    this.maxSubtreeDepth = Math.max(this.left.maxSubtreeDepth, this.right.maxSubtreeDepth) + 1;
+    this.range = NumericRange.createFromInterval(this.left.range.from, this.right.range.toInclusive);
+    this.count = this.left.count + this.right.count;
+  }
+
+  static fromIndexNodesAndBalance(r0: IndexNode, r1: IndexNode) {
+    if (r0.range.from > r1.range.toInclusive + 1) {
```

Comment:
NumericRange.toInclusive seems to consistently be used with a '+ 1', maybe just have normal convection and not use 'inclusive to' and rather 'exclusive to'

Reply:
Skipped this because it just caused a lot of problems. I'm too stupid - sorry.